plugins/parsers/csv/parser.go

```diff
@@ -0,0 +1,181 @@
+		}
+		csvReader.Comma = runeStr[0]
+	}
+	if p.Comment != "" {
+		runeStr := []rune(p.Comment)
+		if len(runeStr) > 1 {
+			return csvReader, fmt.Errorf("comment must be a single character, got: %s", p.Comment)
+		}
+		csvReader.Comment = runeStr[0]
+	}
+	csvReader.TrimLeadingSpace = p.TrimSpace
+	return csvReader, nil
+}
+
+func (p *CSVParser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	r := bytes.NewReader(buf)
+	csvReader, err := p.compile(r)
+	if err != nil {
+		return nil, err
+	}
+	// if there is a header and nothing in DataColumns
+	// set DataColumns to names extracted from the header
+	if p.Header && len(p.DataColumns) == 0 {
+		header, err := csvReader.Read()
+		if err != nil {
+			return nil, err
+		}
+		p.DataColumns = header
+
+	} else if p.Header {
+		// if there is a header and DataColumns is specified, just skip header
+		csvReader.Read()
+
+	} else if !p.Header && len(p.DataColumns) == 0 {
+		// if there is no header and no DataColumns, that's an error
+		return nil, fmt.Errorf("there must be a header if `csv_data_columns` is not specified")
+	}
+
+	table, err := csvReader.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, record := range table {
+		m, err := p.parseRecord(record)
+		if err != nil {
+			return metrics, err
+		}
+		metrics = append(metrics, m)
+	}
+	return metrics, nil
+}
+
+// ParseLine does not use any information in header and assumes DataColumns is set
+func (p *CSVParser) ParseLine(line string) (telegraf.Metric, error) {
+	r := bytes.NewReader([]byte(line))
+	csvReader, err := p.compile(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// if there is nothing in DataColumns, ParseLine will fail
+	if len(p.DataColumns) == 0 {
+		return nil, fmt.Errorf("[parsers.csv] data columns must be specified")
+	}
+
+	record, err := csvReader.Read()
+	if err != nil {
+		return nil, err
+	}
+	m, err := p.parseRecord(record)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (p *CSVParser) parseRecord(record []string) (telegraf.Metric, error) {
+	recordFields := make(map[string]string)
+	tags := make(map[string]string)
+	fields := make(map[string]interface{})
+	for i, fieldName := range p.DataColumns {
+		recordFields[fieldName] = record[i]
+	}
+
+	// add default tags
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	for _, tagName := range p.TagColumns {
+		if recordFields[tagName] == "" {
+			return nil, fmt.Errorf("could not find field: %v", tagName)
+		}
+		tags[tagName] = recordFields[tagName]
+	}
+
+	for _, fieldName := range p.FieldColumns {
+		value, ok := recordFields[fieldName]
+		if !ok {
+			return nil, fmt.Errorf("could not find field: %v", fieldName)
+		}
+
+		// attempt type conversions
+		if iValue, err := strconv.Atoi(value); err == nil {
+			fields[fieldName] = iValue
+		} else if fValue, err := strconv.ParseFloat(value, 64); err == nil {
```

Comment:
This will require all floats to have a decimal part to avoid type mismatch errors.  @goller Is this going to work for us in the future?